### ISSUE TITLE
Update output `final` logic

### ISSUE THREAD

**Issue by pgrete** [parthenon-hpc-lab/parthenon#752]
the `final` output names typically only make sense when the simulation stopped before reaching a limit set by the input file (such as `nlim` or `tlim`), e.g., by reaching a wallclock limit or being interrupted by a signal.
Thus, we could only use `final` in that case so that the last (intended) output automatically also gets a proper id.

**Comment by c-prather** [comment]
I like this.  It was a handful of lines to put together a prototype and it feels like as a "breaking" change it should go in with the file format breaks for next release.

Should hitting nsteps be a "final" or numbered dump?  I'm inclined to say "final," since I find it much more likely that it's off-cadence with other dumps spaced evenly in simulation time.

@Yurlungur may have opinions. Not sure if other downstreams rely on the current behavior, I certainly don't.

**Comment by pgrete** [comment]
> Should hitting nsteps be a "final" or numbered dump? I'm inclined to say "final," since I find it much more likely that it's off-cadence with other dumps spaced evenly in simulation time.

I'm also inclined to say final for the same reason.
The argument could also go along "outputs are determined by a `dt` so everything with this `dt` should get numbers, the last output triggered by anything else (nlim, wallclock, signal/file) should be `final`"
This logic, of course, may change if we allow output based on `dn` for which there are use cases (but I don't need them yet).

**Comment by tagardiner** [comment]
Just my 2 cents - I've only ever used numbered outputs.  It may just be me, but when I choose a dt for output, I typically choose it to be an integer fraction of the simulation run time, i.e. (tend - tstart) = n dt for n an integer.  But, as most things go with simulations, "don't give shotguns to ducks" as John Hawley would say.  This also makes sense to me from the point of restarting a simulation to run longer.

I do still use outputs incremented on time-step from time to time when I want too see what's happening with the solution or refinement or ...  This is especially helpful when restarting around a problematic solution time.

Can we make this question of numbered final output an option?

**Comment by brryan** [comment]
Using `final` only when the simulation is interrupted makes sense to me. We recently merged in #729 which makes the `final` labeling entirely optional (disabled by the user by setting `use_final_label = false` in any output block in the input parameter file). Maybe we could extend the `use_final_label` parameter to accept `always`, `never`, and `only_on_interrupt` options? My preference would be to make `only_on_interrupt` the default.

**Comment by c-prather** [comment]
Agreed re: option.  An expanded naming option should be easy enough to put into the logic, and as a bonus I feel less bad not checking with every downstream before posting the PR.  I can see someone maybe running a sequence of simulations chained from the output of the last sim: starting based on the last numbered dump requires some searching, where 'final' is more reliable.

Why did I not encounter the use_final_label logic in prototyping this I wonder?  It looks like the final dump logic is kind of split across the driver and output machinery, where it should really be restricted to output with comprehensive flags for the types of things that the output machinery can do.  I'll try to address that I think -- the less we put in Driver, the more people not subclassing can still benefit from improvements.

Thinking about this more, the current framework lacks flexibility badly.  Say for example:
1. I have a massive ~1TB output I do once or twice a sim.  That probably shouldn't be triggered along with everything else at the end
2. I want to simplify my code by assuming that history file lines are always written exactly on cadence.  Now I have do deal with stray lines whenever the scheduler ended my jobs.
3. I want to write a restart file at the simulation end, but I really don't need or want a dump file, regardless of what it's named.
4. I want to write a restart file in place of every Xth dump, because they contain the same information anyway and I want to save space while being able to re-run any section of the simulation with e.g. new physics.
5. I'd like to output with custom strings, or with integers following time rather than sequence.

If there's a backwards-compatibility option I think that takes away some of the time pressure to put all the output-breaking changes in one release, so if there are no objections let me casually reserve this for "sometime in the next month unless somebody needs it soon" and see if the resulting design can't cover at least a few more use cases.

**Comment by Yurlungur** [comment]
Using `final` only if the sim is interrupted also makes sense to me. I think that's my preferred modality. But I don't see a big reason to give up the flexibility of setting this in the input deck.

**Comment by pgrete** [comment]
Agreed. I like @brryan suggestion in terms of input deck options and default vals.

**Comment by c-prather** [comment]
Going to take a look at this soon I swear...
One additional option recently came up in multi-scale simulations: we should also have `dn = int` output options, for outputting every `n` steps.

**Comment by Yurlungur** [comment]
> One additional option recently came up in multi-scale simulations: we should also have dn = int output options, for outputting every n steps.

+1 that would be a nice thing to have

**Comment by pgrete** [comment]
+2 ;)

**Comment by BenWibking** [comment]
This would be extremely useful to have.
